hs-tendermint/src/Network/Tendermint.hs

```diff
@@ -0,0 +1,271 @@
+data SimpleProof = SimpleProof
+  { simpleProofTotal    :: Int
+  , simpleProofIndex    :: Int
+  , simpleProofLeafHash :: HexString
+  , simpleProofAunts    :: [HexString]
+  } deriving (Eq, Show, Generic)
+instance FromJSON SimpleProof where
+  parseJSON = genericParseJSON $ defaultRPCOptions "simpleProof"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/block_meta.go#L4
+data BlockMeta = BlockMeta
+  { blockMetaBlockId :: FieldTypes.BlockID
+  , blockMetaHeader  :: FieldTypes.Header
+  } deriving (Eq, Show, Generic)
+instance FromJSON BlockMeta where
+  parseJSON = genericParseJSON $ defaultRPCOptions "blockMeta"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/block.go#L36
+data Block = Block
+  { blockHeader     :: FieldTypes.Header
+  , blockData       :: Data
+  , blockEvidence   :: EvidenceData
+  , blockLastCommit :: Maybe Commit
+  } deriving (Eq, Show, Generic)
+instance FromJSON Block where
+  parseJSON = genericParseJSON $ defaultRPCOptions "block"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/block.go#L774
+data Data = Data
+  { dataTxs :: [Tx]
+  } deriving (Eq, Show, Generic)
+instance FromJSON Data where
+  parseJSON = genericParseJSON $ defaultRPCOptions "data"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/block.go#L819~
+data EvidenceData = EvidenceData
+  { evidenceDataEvidence :: EvidenceList
+  } deriving (Eq, Show, Generic)
+instance FromJSON EvidenceData where
+  parseJSON = genericParseJSON $ defaultRPCOptions "evidenceData"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/evidence.go#L278
+type EvidenceList = [FieldTypes.Evidence]
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/block.go#L488
+data Commit = Commit
+  { commitBlockId    :: FieldTypes.BlockID
+  , commitPrecommits :: [Vote]
+  } deriving (Eq, Show, Generic)
+instance FromJSON Commit where
+  parseJSON = genericParseJSON $ defaultRPCOptions "commit"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/vote.go#L51
+data Vote = Vote
+  { voteType             :: SignedMsgType
+  , voteHeight           :: Int64
+  , voteRound            :: Int
+  , voteBlockId          :: FieldTypes.BlockID
+  , voteTimestamp        :: FieldTypes.Timestamp
+  , voteValidatorAddress :: HexString
+  , voteValidatorIndex   :: Int
+  , voteSignature        :: HexString
+  } deriving (Eq, Show, Generic)
+instance FromJSON Vote where
+  parseJSON = genericParseJSON $ defaultRPCOptions "vote"
+
+-- https://github.com/tendermint/tendermint/blob/14fa8007738c51fe0951544a63eeea759679d495/types/tx.go#L19
+type Tx = HexString
```

Comment:
In general this kind of newtyping doesn't really add very much value, I think it's fine to leave it as `HexString` since that's become synonymous with `ByteString`

Reply:
It's not a newtype, it's type alias. I tried to sort of port types from go 1 to 1. I think it would make it easy to debug stuff once things go wrong. do you think otherwise?

Reply:
sorry I meant aliasing . In general I think type aliasing does well when you’re able to supply type some type variables with defaults (like halogen does this I’m pretty sure), but usually in this case I’m not for it because you don’t gain type safety and there will be places where we refer to Txs by just HexString, like in the other packages. You can get pretty far with just using haddock , see the bottom of this page https://www.haskell.org/haddock/doc/html/ch03s02.html